collector/fibrechannel_linux.go

```diff
@@ -0,0 +1,149 @@
+		"invalid_crc_total":              "Invalid Cyclic Redundancy Check count",
+		"nos_total":                      "Number Not_Operational Primitive Sequence received by host port",
+		"fcp_packet_aborts_total":        "Number of aborted packets",
+		"rx_words_total":                 "Number of words received by host port",
+		"tx_frames_total":                "Number of frames transmitted by host port",
+		"link_failure_total":             "Number of times the host port link has failed",
+		"name":                           "Name of Fibre Channel HBA",
+		"speed":                          "Current operating speed",
+		"port_state":                     "Current port state",
+		"port_type":                      "Port type, what the port is connected to",
+		"symbolic_name":                  "Symoblic Name",
+		"node_name":                      "Node Name as hexadecimal string",
+		"port_id":                        "Port ID as string",
+		"port_name":                      "Port Name as hexadecimal string",
+		"fabric_name":                    "Fabric Name; 0 if PTP",
+		"dev_loss_tmo":                   "Device Loss Timeout in seconds",
+		"supported_classes":              "The FC classes supported",
+		"supported_speeds":               "The FC speeds supported",
+	}
+
+	i.metricDescs = make(map[string]*prometheus.Desc)
+	i.subsystem = "fibrechannel"
+
+	for metricName, description := range descriptions {
+		i.metricDescs[metricName] = prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, i.subsystem, metricName),
+			description,
+			[]string{"fc_host"},
+			nil,
+		)
+	}
+
+	return &i, nil
+}
+
+func (c *fibrechannelCollector) pushMetric(ch chan<- prometheus.Metric, name string, value uint64, host string, valueType prometheus.ValueType) {
+	ch <- prometheus.MustNewConstMetric(c.metricDescs[name], valueType, float64(value), host)
+}
+
+func (c *fibrechannelCollector) pushCounter(ch chan<- prometheus.Metric, name string, value uint64, host string) {
+	// Don't push counters that aren't implemented (a counter equal to maxUint64 is unimplemented by the HBA firmware)
+	if value != maxUint64 {
+		c.pushMetric(ch, name, value, host, prometheus.CounterValue)
+	}
+}
+
+func (c *fibrechannelCollector) Update(ch chan<- prometheus.Metric) error {
+	hosts, err := c.fs.FibreChannelClass()
+	if err != nil {
+		if os.IsNotExist(err) {
+			level.Debug(c.logger).Log("msg", "fibrechannel statistics not found, skipping")
+			return ErrNoData
+		}
+		return fmt.Errorf("error obtaining FibreChannel class info: %s", err)
+	}
+
+	for _, host := range hosts {
+		infoDesc := prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, c.subsystem, "info"),
+			"Non-numeric data from /sys/class/fc_host/<host>, value is always 1.",
+			[]string{"host", "speed", "port_state", "port_type", "port_id", "port_name", "fabric_name", "symbolic_name", "supported_classes", "supported_speeds", "dev_loss_tmo"},
```

Comment:
Shouldn't this also be `fc_hsot`?

```suggestion
			[]string{"fc_host", "speed", "port_state", "port_type", "port_id", "port_name", "fabric_name", "symbolic_name", "supported_classes", "supported_speeds", "dev_loss_tmo"},
```

Reply:
well that makes sense; been too long since I originally wrote this. :)